rclcpp/src/rclcpp/publisher.cpp

```diff
@@ -231,6 +231,19 @@ PublisherBase::setup_intra_process(
   IntraProcessManagerSharedPtr ipm,
   const rcl_publisher_options_t & intra_process_options)
 {
+  // Skip intraprocess configuration if the "durability" qos policy is not "volatile".
+  // TODO(ivanpauno):
```

Comment:
It would be good, in general, to have methods to check what is the real value of RMW_QOS_*_SYSTEM_DEFAULT. Or maybe, just delete all these defines and only support explicit configurations. Having unknown default values doesn't seem like a good idea.

Reply:
Correct snippet of the TODO:
https://github.com/ros2/rclcpp/blob/c31b273c1fc8802dcdd293266b1b5212ec9583ef/rclcpp/src/rclcpp/publisher.cpp#L235-L246

Reply:
Personally I agree with you that we should avoid implicit settings, which is what the `SYSTEM_DEFAULT` gives you, but there's a use case that some people want supported, which allows you to configure QoS settings from external XML files. That is what this allows for, so I don't know if we could realistically get rid of those.

We definitely should have a way to check the effective QoS settings after the publisher is created.

Also, I think this (re)raises the issue that `use_intra_process_comms` needs to be an option on publishers/subscriptions, and not on the node itself. Otherwise you'd be unable to use intra process _and_ have a non-volatile durability on one of your publishers.

Once we have that, I'd say it should be an error (not a warning) to have both intra process enabled and transient local durability on the same publisher/subscription.

Reply:
I agree about having use_intra_process_comms option at publish/subscription level. Switching the warning to an error after that also sounds good.

Reply:
I still find this point unresolved, until we have the ability to know what the durability is when `SYSTEM_DEFAULT` is in use and until we have the ability to control which publishers and subscriptions have intra-process enabled.

---

To be clear about the first point, I think we need a way to inspect a publisher or subscription after creation to see which durability it is using. Something like `rcl_ret_t rcl_publisher_get_actual_qos(const rcl_publisher_t *, rmw_qos_profile *)`, which may need an `rmw` equivalent.

I know this increases the scope of this issue considerably, but without it the warning is only going to work part of the time and would produce very subtle bugs for the user.

The alternative is an even bigger scope increase, which is to actually support transient-local durability correctly when intra-process is enabled.

---

For the second point, what I imagine is that the creation publishers and subscriptions have a three-way option which is "explicitly enable intra-process", "explicitly disable intra-process", or "default". Where "default" would be based on the intra-process setting at the node level.

I think this is important to do now (again I know it increases the scope of this issue), because otherwise you have to use a warning rather than an error to let the user know that they have conflicting configurations. Without it, you also have scenarios which are impossible to achieve, i.e. using intra-process and transient-local durability at the same time but on different publishers and/or subscriptions.